### ISSUE TITLE
filterColumn with laravel polymorphic relationships

### ISSUE THREAD

**Issue by Esperancia** [yajra/laravel-datatables#2291]
### Summary of problem or feature request

I am using Laravel Yajra Datatables with polymorphic relations. the specific trouble is that I can neither filter the table nor sort properly because I populated it with different tables values.

The table A is "morphable" to tables B and C. So it has column

> morph_type and morph_id.

but also other different columns as

B.name , B.surname, C.lastname, C.firstname

Because filter does not work, I tried to implement Datatables filterColumn API. But I do not really know the right way to this query. Here is my sql (Postgresql) :

### Code snippet of problem

```php
$query = A::with('morphable')->select('*');
return Datatables::of($query)
        ->addColumn('code', function($row) {
            return $row->morphable->code_pap;
        })
        ->addColumn('name', function($row) {
            if($row->morphable->name !== null){
                return $row->morphable->name;
            } else {
                return $row->morphable->lastname;
            }
        })
        ->addColumn('surname', function($row) {
            if($row->morphable->surname !== null){
                return $row->morphable->surname;
            } else {
                return $row->morphable->firstname;
            }
        })
        ->filterColumn('name', function($query, $keyword) {
            //I check $query it retruns "select * from A;" only.
            // redefining $query does not change anything. I always get same query.
            $sqlB = "CASE WHEN exists (select 1 from B where B.name like ?)";
            $sqlC = "CASE WHEN exists (select 1 from C where C.lastname like ?)";
            $query->whereRaw($sqlB, ["%{$keyword}%"])->orWhereRaw($sqlC, ["%{$keyword}%"]);
        }) //same thing for firstname and surname ;
```

### System details

- ubuntu
- PHP 7.4.0
- laravel 6
- yajra/laravel-datatables-oracle: "^9.7"

I check $query inside filterColum it returns "select * from A;" only. Redefining $query does not change anything. I always get same query.

**Comment by Esperancia** [comment]
I find a way to filter by keyword, but sorting added column not working seems to not working properly...
`
     ...

        ->filterColumn('name', function($query, $keyword) {
            $query->whereHasMorph('morphable', [B::class, C::class],
            function ($q, $type) use($keyword) {
                if ($type === B::class) {
                    $q->whereRaw("LOWER(name) LIKE ?", "%".strtolower($keyword)."%"); 
                } //sql with lowercase comparison just to show it
                if ($type === C::class) {
                    $q->whereRaw("UPPER(lastname) LIKE ?", "%".strtoupper($keyword)."%"); 
                } //sql with uppercase comparison just to show it
            });
        }); //same for the next field
     ...

`

**Comment by RyanPriceDotCa** [comment]
In case it helps, here is an example for using filterColumn with a polymorphic relationship from a project I am working on.  In this example, the key tables in play are:

1. Extension (this has the morphTo relationship)
2. Person (morphMany relationship)
3. Service (morphMany relationship)

An extension can belong to a person or service. 

**From Extension Model**
```php
/**
     * Get the owning extensionable table
     */
    public function extensionable() {
        return $this->morphTo(null, null, null, 'directoryid');
    }
```
**From Person and Service Model (relationship looks the same in both)**
```php
    public function extensions()
    {
        return $this->morphMany('App\Extension', 'extensionable', null, null, 'directoryid');
    }
```
NOTE: *I use a different id for the relationship, you probably don't require all those parameters if you followed laravel naming conventions.

Here are the relevant pieces of the DataTable class: 

```php
public function dataTable($query)
    {
        return datatables()
            ->eloquent($query)
            ->editColumn('extensionable.name', function (Extension $extension) {
                return $extension->extensionable->getFullNameAttribute();
            })
            ->filterColumn('extensionable.name', function ($query, $keyword) {
                $extensions = Extension::whereHasMorph('extensionable', [Person::class, Service::class], function ($subquery, $type) use ($keyword) {
                    if ($type === Person::class) {
                        $subquery->where('given_name', 'like', '%'.$keyword.'%')
                        ->orWhere('surname', 'like', '%'.$keyword.'%')
                        ->orWhere(DB::raw('CONCAT(given_name, " ", surname)'), 'like', '%' . $keyword . '%');
                    }

                    if ($type === Service::class) {
                        $subquery->where('primary_name', 'like', '%'.$keyword.'%')
                        ->orWhere('secondary_name', 'like', '%'.$keyword.'%')
                        ->orWhere(DB::raw('CONCAT(primary_name, " ", secondary_name)'), 'like', '%' . $keyword . '%');
                    }
                })->get()->pluck('id')->toArray();

                $query->whereIn('id', $extensions);
            })
            ->editColumn('action', function (Extension $extension) {
                return '<a href="' . route('extension.edit', $extension) . '" class="btn btn-xs btn-primary"><i class="glyphicon glyphicon-edit"></i> Edit</a>';
            });
    }

   /**
     * Get columns.
     *
     * @return array
     */
    protected function getColumns()
    {
        return [
            //showing only the column that matters in this example
            Column::make('extensionable.name')->title('Belongs To')->sortable(false),
        ];
    }
```

`whereHasMorph` is only available in laravel v5.8.27 and above.  See https://laravel-news.com/laravel-5-8-27 for details

Hope this helps someone.
